app/controllers/transactions_controller.rb

```diff
@@ -1,15 +1,104 @@
 class TransactionsController < ApplicationController
 
-  before_filter do |controller|
+  before_filter only: [:show] do |controller|
     controller.ensure_logged_in t("layouts.notifications.you_must_log_in_to_view_your_inbox")
   end
 
-  before_filter :only => [ :index, :received ] do |controller|
-    controller.ensure_authorized t("layouts.notifications.you_are_not_authorized_to_view_this_content")
+  before_filter do |controller|
+    controller.ensure_logged_in t("layouts.notifications.you_must_log_in_to_do_a_transaction")
   end
 
   MessageForm = Form::Message
 
+  TransactionForm = EntityUtils.define_builder(
+    [:listing_id, :fixnum, :to_integer, :mandatory],
+    [:message, :string],
+    [:quantity, :fixnum, :to_integer, default: 1],
+    [:start_on, transform_with: ->(v) { Maybe(v).map { |d| TransactionViewUtils.parse_booking_date(d) }.or_else(nil) } ],
+    [:end_on, transform_with: ->(v) { Maybe(v).map { |d| TransactionViewUtils.parse_booking_date(d) }.or_else(nil) } ]
+  )
+
+  def new
+    Result.all(
+      ->() {
+        fetch_data(params[:listing_id])
+      },
+      ->((listing_id, listing_model)) {
+        ensure_can_start_transactions(listing_model: listing_model, current_user: @current_user, current_community: @current_community)
+      }
+    ).on_success { |((listing_id, listing_model, author_model, process, gateway))|
+      booking = listing_model.unit_type == :day
+
+      transaction_params = {listing_id: listing_model.id}.merge(params.slice(:start_on, :end_on, :quantity))
+
+      case [process[:process], gateway, booking]
+      when matches([:none])
```

Comment:
This was just moved from one place to another, so it's ok as it is.
(But I find it a bit too heavy to follow)